client/src/client/es/client/client.ts

```
@@ -0,0 +1,190 @@
+
+  constructor(errorCode: number, message: string, errors?: ApiErrors) {
+    super(message);
+
+    this.errorCode = errorCode;
+    this.errors = errors;
+  }
+}
+
+export function createApiClient(
+  apiHost: string,
+  authProvider: AuthProvider,
+): ApiClient {
+  async function request<P, R>(
+    method: 'GET' | 'POST' | 'DELETE' | 'PUT' | 'PATCH',
+    endpoint: string,
+    authRequired = true,
+    params?: P,
+  ): Promise<ApiResponse<R>> {
+    const url = new URL(`${apiHost}${endpoint}`.replaceAll('//', '/'));
+    const headers = new Headers();
+
+    headers.set('Content-type', 'application/json');
+    headers.set('Accept', 'application/json');
+
+    const requestInit: RequestInit = {
+      method,
+      mode: 'cors',
+      cache: 'no-cache',
+      headers,
+      redirect: 'follow',
+    };
+
+    if (['POST', 'PATCH', 'PUT'].includes(method)) {
+      const body = {
+        ...params,
+      };
+
+      requestInit.body = JSON.stringify(body);
+    }
+
+    if (method === 'GET') {
+      const searchParams = {
+        ...params,
+      };
+      url.search = new URLSearchParams(searchParams).toString();
+    }
+
+    const request = await authProvider(
+      new Request(url.toString(), requestInit),
+      authRequired,
+    );
+
+    let response: Response;
+
+    try {
+      response = await fetch(request);
+    } catch (e) {
+      throw new ApiClientError(500026, 'Network error');
+    }
+
+    switch (response.status) {
+      case 200:
+        try {
+          return {
+            status: response.status,
+            body: (await response.json()) as R,
+            headers: response.headers,
+          } as ApiBodyResponse<R>;
+        } catch (e) {
+          throw new ApiClientError(response.status, 'Invalid JSON response');
+        }
+      case 204:
+        return {
+          status: response.status,
+          headers: response.headers,
+        } as ApiEmptyResponse;
+      default:
+        return {
```

Maybe consider throwing an error instead of returning object here? Handling should be much easier then